CountingSort/app_resources/scatter_shader.comp.hlsl

```diff
@@ -14,16 +18,34 @@ uint32_t3 nbl::hlsl::glsl::gl_WorkGroupSize()
 void main(uint32_t3 ID : SV_GroupThreadID, uint32_t3 GroupID : SV_GroupID)
 {
     uint32_t tid = nbl::hlsl::workgroup::SubgroupContiguousIndex();
+
+    [unroll]
+    for (int i = 0; i < BucketsPerThread; i++)
+        sdata[BucketsPerThread * tid + i] = 0;
     uint32_t index = (nbl::hlsl::glsl::gl_WorkGroupID().x * WorkgroupSize + tid) * pushConstants.elementsPerWT;
 
     nbl::hlsl::glsl::barrier();
     
+    [unroll]
     for (int i = 0; i < pushConstants.elementsPerWT; i++)
     {
         if (index + i >= pushConstants.dataElementCount)
             break;
         uint32_t value = vk::RawBufferLoad(pushConstants.inputAddress + sizeof(uint32_t) * (index + i));
-        uint32_t address = nbl::hlsl::glsl::atomicAdd(scratch[value - pushConstants.minimum], (uint32_t) -1);
-        vk::RawBufferStore(pushConstants.outputAddress + sizeof(uint32_t) * (address - 1), value);
+        nbl::hlsl::glsl::atomicAdd(sdata[value - pushConstants.minimum], (uint32_t) 1);
+    }
+
+    nbl::hlsl::glsl::barrier();
+
+    [unroll]
+    for (int i = 0; i < BucketsPerThread; i++)
+    {
+        uint32_t count = sdata[WorkgroupSize * i + tid];
+        sdata[WorkgroupSize * i + tid] = nbl::hlsl::glsl::atomicAdd(scratch[WorkgroupSize * i + tid], sdata[WorkgroupSize * i + tid]);
+        [unroll]
+        for (int j = 0; j < count; j++)
+        {
+            vk::RawBufferStore(pushConstants.outputAddress + sizeof(uint32_t) * (sdata[WorkgroupSize * i + tid] + j), WorkgroupSize * i + tid);
+        }
```

Comment:
this is bad work-balancing, you still want 1 thread to perform `elementsPerWT` stores

I know you're trying to avoid a second `vk::RawBufferLoad` but you'd usually sort Key,Value SoA pairs, so they histogramming passes would only ready Key, while the final scatter here would touch `Value` instead

you need a fourth loop which does Load/Store while atomicAdd(-1) on the `sdata` to figure out the destination index

Reply:
For the Key,Value pairs, are we supposed to sort by Key or Value? I thought we were supposed to sort by Value? Other than that, we do that now

Reply:
You sort by Key obviously, its associated Value just moves with it.

So if key originally at offset i ends up at offset j, the value should too.

This is why you only histogram keys, but scatter both.

Reply:
Done